amazon-kinesis-client-multilang/src/main/java/software/amazon/kinesis/multilang/config/AwsCredentialsProviderPropertyValueDecoder.java

```diff
@@ -153,28 +162,43 @@ private static List<String> getPossibleFullClassNames(final String provider) {
     }
 
     @FunctionalInterface
-    private interface CredentialsProviderConstructor<T extends AWSCredentialsProvider> {
+    private interface CredentialsProviderConstructor<T extends AwsCredentialsProvider> {
         T construct()
                 throws IllegalAccessException, InstantiationException, InvocationTargetException, NoSuchMethodException;
     }
 
     /**
-     * Attempts to construct an {@link AWSCredentialsProvider}.
+     * Attempts to construct an {@link AwsCredentialsProvider}.
      *
      * @param providerName Raw, unmodified provider name. Should there be an
-     *      Exeception during construction, this parameter will be logged.
+     *      Exception during construction, this parameter will be logged.
      * @param constructor supplier-like function that will perform the construction
+     * @param clazz the class to attempt to construct
      * @return the constructed provider, if successful; otherwise, null
      *
      * @param <T> type of the CredentialsProvider to construct
      */
-    private static <T extends AWSCredentialsProvider> T constructProvider(
-            final String providerName, final CredentialsProviderConstructor<T> constructor) {
+    private static <T extends AwsCredentialsProvider> T constructProvider(
+            final String providerName,
+            final CredentialsProviderConstructor<T> constructor,
+            Class<? extends AwsCredentialsProvider> clazz) {
         try {
+            // Try to create an instance using the given constructor
             return constructor.construct();
+        } catch (InstantiationException e) {
```

Comment:
This is sloppy and needs a second look. The issue here is that we relied on this provider constructor to create any provider class that is passed to this function.

However, in our tests, we use `software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider`. This used to be invoked like `new DefaultCredentialsProvider` but now is constructed with `DefaultCredentialsProvider.create()` - https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/migration-client-credentials.html

I tried to generalize this to work with both, but its quite sloppy.

Reply:
Users can still create classes that implement the v2 `AwsCredentialsProvider` that use a default constructor `clazz::newInstance` instead of a `builder`/`create`.

For instance, we create a basic [AwsCredentialsProvider that relies on a default constructor in our testing](https://github.com/ethkatnic/amazon-kinesis-client/blob/444a4ddf9aa984865db51c702dc78e505d2954f7/amazon-kinesis-client-multilang/src/test/java/software/amazon/kinesis/multilang/config/KinesisClientLibConfiguratorTest.java#L528). 

Since this is still a valid way to create v2 providers, I think this code should support both a `.create()` method and a default constructor method of instantiating providers.

Reply:
Another thing to consider is that we currently support passing args to default constructors through the config like `ConstructorCredentialsProvider|arg1|arg2 `. This eventually calls a constructor like `provider = ConstructorCredentialsProvider(arg1, arg2)`, allowing the user to pass in a provider string with constructor args.

I'm not sure how we would accomplish this with providers that use builders. For instance, if a user wanted to create a DefaultCredentialsProvider with `profileName = "test-profile"` like `DefaultCredentialsProvider.builder().profileName("Test").build()`, the code currently doesn't support this. My above method would just take `DefaultCredentialsProvider|arg1|arg2` and invoke DefaultCredentialsProvider.create() to create a default instance of the class. 

Do you think this is something we need to support?

Reply:
Yes, we should come up with a way that's easy for customer to use the default credential providers without having to peek into the multilang code or documentation to pick the providers. Let's iterate on this